app/ui/browser/lib/user-agent.js

```diff
@@ -14,28 +14,51 @@ import request from './request';
 import * as endpoints from '../../../shared/constants/endpoints';
 import { getPageById } from '../selectors/index';
 
-export function createPage({ url, session, page }) {
-  return request(`/pages/${encodeURIComponent(url)}`, {
-    method: 'POST',
-    json: {
-      session,
-      page,
-    }
-  });
+/**
+ * A mapping from a page id to a promise that
+ * will resolve to a session id. When we initiate the session,
+ * we immediately store a promise that will resolve to the session id
+ * so that future requests using the page id can hook into the initial promise
+ * even before we have the session.
+ */
+const PAGE_TO_SESSION = new Map();
+
+async function waitForSession(page) {
+  // If we have a promise to the session for this page,
+  // that means we requested a new session, so use it.
+  if (PAGE_TO_SESSION.has(page.id)) {
+    return PAGE_TO_SESSION.get(page.id);
+  }
+  // If we don't have a session promise saved, we may have used a
```

Comment:
Moving this to first -- fixing up tests, and since we're not running a UA service in unit tests, this will actually work once flipping this order.